lib/eth/types/faceted_numeric.rb

```diff
@@ -0,0 +1,124 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# -*- encoding : ascii-8bit -*-
+
+# Provides the {Eth} module.
+module Eth
+  
+  # Provides the `Types` module supporting various data types.
+  module Types
+
+    # Provides a faceted, numeric data type to help alleviate
+    #   the pain of converting from a keccak256 binary string -> hex -> prefixed hex,
+    #   vice versa, and other conversions.
+    #
+    # @see {https://github.com/q9f/secp256k1.cr/blob/master/src/secp256k1/num.cr} Num from secp256k1.cr
+    # @see #{Eth::Util} is used heavily throughout
+    # @example Pass in a 0x-prefixed hex string, and coerce to keccak256 binary
+    #   FacetedNumeric.new('0xDEADBeeF').to_binary
+    #   FacetedNumeric['0xDEADBeeF'].to_keccak256
+    # @raise [ArgumentError] raised if input is a String but
+    #   isn't hex or bytes, or if input is an Array but not
+    #   a list of ascii-8bit numbers
+    class FacetedNumeric < Numeric
+      include Util
+
+      # raw input
+      attr_reader :input
+
+      # value in base 10
+      attr_reader :integer
+
+      # binary (keccak256)
+      # @see {Eth::Util.keccak256}
+      attr_reader :binary
+
+      # hexadecimal
+      attr_reader :hex
+
+      # 0x-prefixed hexadecimal
+      attr_reader :prefixed_hex
+
+      # zero-padded hexadecimal
+      attr_reader :zpadded_hex
+
+      # zero-padded binary
+      attr_reader :zpadded_binary
+
+      # @param [nil, Integer, String<Bytes>, String<Hex>]
+      def initialize(input = nil)
+        @input = input
+        coerce!
+      end
+      #alias [] initialize
+
+      # Eagerly set all value types.
+      def coerce!
+        case input
+        when Integer
+          @integer = input
+          @hex = integer.to_s(16)
+          @binary = hex_to_bin(hex)
+        when method(:is_bytes?)
+          @binary = input
+          @hex = bin_to_hex(binary)
+          @integer = Integer(hex, 16)
+        when method(&:is_hex?)
+          @hex = remove_hex_prefix(input)
+          @integer = Integer(input, 16)
+          @binary = hex_to_bin(hex)
+        when nil
+          generate_key
+        else
+          raise ArgumentError, "Invalid input: #{input}"
+        end
+      end
+
+      private def generate_key
+        @hex = SecureRandom.hex(32)
+        @integer = Integer(hex, 16)
+        @binary = hex_to_bin(hex)
+      end
+
+      def to_i
+        integer
+      end
+
+      def to_hex
+        hex 
+      end
+
+      def to_binary
+        binary
+      end
+
+      def to_keccak256
+        keccak256(to_prefixed_hex)
+      end
+
+      def to_prefixed_hex
+        @prefixed_hex ||= prefix_hex(hex)
+      end
+
+      def to_zpadded_hex
```

Comment:
Ideally, you want to allow specifying how many zeros you want to pad.

Reply:
Fixed